pkg/oci/routes.go

```diff
@@ -0,0 +1,291 @@
+	}
+
+	ipOcid, err := cp.client.Networking().GetOCIDFromIP(ctx, *vnic.PrivateIp, *vnic.SubnetId)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	glog.V(6).Infof("ipOcid: ", ipOcid)
+
+	for routeTableID := range routeTableIds {
+		routeTable, err := cp.client.Networking().GetRouteTable(ctx, routeTableID)
+		if err != nil {
+			return errors.WithStack(err)
+		}
+		for _, oRoute := range routeTable.RouteRules {
+			if *oRoute.CidrBlock == route.DestinationCIDR {
+				if *oRoute.NetworkEntityId == ipOcid {
+					// skipping existing route table entry
+					glog.V(6).Info("Skipping existing route entry: ", route.DestinationCIDR, " -> ", ipOcid)
+					continue
+				} else {
+					oRoute.NetworkEntityId = &ipOcid
+					glog.V(6).Info("Update existing route entry: ", route.DestinationCIDR, " -> ", ipOcid)
+					err := cp.client.Networking().UpdateRouteTable(ctx, routeTableID, routeTable.RouteRules)
+					if err != nil {
+						return errors.WithStack(err)
+					}
+					continue
+				}
+			}
+		}
+
+		// route rule doesn't exist in RouteTable
+		glog.V(6).Info("Add new route entry: ", route.DestinationCIDR, " -> ", ipOcid)
+		routeTable.RouteRules = append(routeTable.RouteRules, core.RouteRule{
+			CidrBlock:       &route.DestinationCIDR,
+			NetworkEntityId: &ipOcid,
+		})
+
+		err = cp.client.Networking().UpdateRouteTable(ctx, routeTableID, routeTable.RouteRules)
+		if err != nil {
+			return errors.WithStack(err)
+		}
+	}
+
+	return nil
+}
+
+// DeleteRoute delete a route entry, but OCI only has Update of the entire RouteTable
+func (cp *CloudProvider) DeleteRoute(ctx context.Context, clusterName string, route *cloudprovider.Route) error {
+	glog.V(6).Infof("Delete Route: %v", route)
+
+	for routeID := range routeTableIds {
+		oRoute, err := cp.client.Networking().GetRouteTable(ctx, routeID)
+		if err != nil {
+			return errors.WithStack(err)
+		}
+
+		dirty := false
+
+		for i, r := range oRoute.RouteRules {
+			if *r.CidrBlock != route.DestinationCIDR {
+				continue
+			}
+
+			dirty = true
+
+			glog.V(6).Infof("Deleting Route: %s %s", route.DestinationCIDR, route.TargetNode)
+			oRoute.RouteRules = append(oRoute.RouteRules[:i], oRoute.RouteRules[i+1:]...)
+		}
+
+		if dirty {
+			glog.V(6).Infof("Updating Route: %s %v", routeID, oRoute.RouteRules)
+			err := cp.client.Networking().UpdateRouteTable(ctx, routeID, oRoute.RouteRules)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+// subnet -> routetable
+// subnets cannot change their route tables, so it's safe to cache
+var subnetIds map[string]string
+
+// routetable -> exists
+// we may see new route tables if we see new subnets
+var routeTableIds map[string]bool
+
+// ListRoutes for OCI, discover all instances and their subnets and those route tables to iterate over
+func (cp *CloudProvider) ListRoutes(ctx context.Context, clusterName string) ([]*cloudprovider.Route, error) {
+	glog.V(6).Info("Listing Routes: ", clusterName)
+
+	if subnetIds == nil {
```

Comment:
This isn't thread safe and the route controller can run with multiple worker goroutines. It's probably better safe than sorry to wrap this in a mutex (can use client-go's cache.Store or sync.Map for somtehing easier)

https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/route/route_controller.go#L151

Reply:
(referring to the shared state maps you're using)